packages/headless/src/controllers/generated-answer/headless-generated-answer.ts

```diff
@@ -0,0 +1,132 @@
+  logOpenGeneratedAnswerSource,
+  logRetryGeneratedAnswer,
+} from '../../features/generated-answer/generated-answer-analytics-actions';
+import {generatedAnswerReducer as generatedAnswer} from '../../features/generated-answer/generated-answer-slice';
+import {GeneratedAnswerState} from '../../features/generated-answer/generated-answer-state';
+import {executeSearch} from '../../features/search/search-actions';
+import {GeneratedAnswerSection} from '../../state/state-sections';
+import {loadReducerError} from '../../utils/errors';
+import {buildController} from '../controller/headless-controller';
+
+export type {GeneratedAnswerState, GeneratedAnswerCitation};
+
+/**
+ * @internal
+ */
+export interface GeneratedAnswer {
+  /**
+   * The state of the GeneratedAnswer controller.
+   */
+  state: GeneratedAnswerState;
+  /**
+   * Re-executes the last query to generate an answer.
+   */
+  retry(): void;
+  /**
+   * Logs a custom event indicating a generated answer is relevant.
+   */
+  likeGeneratedAnswer(): void;
+  /**
+   * Logs a custom event indicating a generated answer is irrelevant.
+   */
+  dislikeGeneratedAnswer(): void;
+  /**
+   * Logs a custom event indicating a cited source link was clicked.
+   */
+  logCitationClick(citation: GeneratedAnswerCitation): void;
+}
+
+/**
+ * @internal
+ */
+export function buildGeneratedAnswer(engine: SearchEngine): GeneratedAnswer {
+  if (!loadGeneratedAnswerReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  const {dispatch} = engine;
+  const controller = buildController(engine);
+  const getState = () => engine.state;
+
+  let source: EventSourcePolyfill;
+  let lastRequestId: string;
+  let lastStreamId: string;
+
+  const setEventSourceRef = (sourceRef: EventSourcePolyfill) => {
+    source = sourceRef;
+  };
+
+  const getIsStreamInProgress = () =>
+    source &&
+    (source.readyState === source.OPEN ||
+      source.readyState === source.CONNECTING);
+
+  const subscribeToSearchRequests = () => {
+    const strictListener = () => {
+      const state = getState();
+      const requestId = state.search.requestId;
+      const streamId =
+        state.search.extendedResults.generativeQuestionAnsweringId;
+
+      if (lastRequestId !== requestId) {
+        lastRequestId = requestId;
+        source?.close();
+        dispatch(resetAnswer());
+      }
+
+      const isStreamInProgress = getIsStreamInProgress();
+      if (!isStreamInProgress && streamId && streamId !== lastStreamId) {
+        lastStreamId = streamId;
+        dispatch(
+          streamAnswer({
+            setEventSourceRef,
+          })
+        );
+      }
+    };
+    return engine.subscribe(strictListener);
+  };
+
+  subscribeToSearchRequests();
+
+  return {
+    ...controller,
+
+    get state() {
+      return getState().generatedAnswer;
+    },
+
+    retry() {
+      dispatch(executeSearch(logRetryGeneratedAnswer()));
+    },
+
+    likeGeneratedAnswer() {
+      dispatch(logLikeGeneratedAnswer());
```

Comment:
There's no state modification other than analytics events on like/dislike ?

Once again, looks very similar to smart snippets, all that's done is to keep in the state if it's liked/disliked so the thumbs up/down can display a state if they are liked/disliked already by the current user/

![image](https://github.com/coveo/ui-kit/assets/2488155/f43fe0c7-08e4-40eb-8cfc-77dd93fded4e)

I'm assuming we're not to that point yet :)